plan/eliminate_projection.go

```diff
@@ -14,50 +14,229 @@
-	}
-	children := make([]Plan, 0, len(p.Children()))
-	for _, child := range p.Children() {
-		children = append(children, EliminateProjection(child.(PhysicalPlan)))
 	}
-	p.SetChildren(children...)
-	return p
-}
-
-// projectionCanBeEliminated checks if a PROJECTION operator can be eliminated.
-// PROJECTION operator can be eliminated when meet the following conditions at the same time:
-// 1. fields of PROJECTION are all columns
-// 2. fields of PROJECTION are just the same as the schema of the child operator (including order, amount, etc.).
-// expressions like following cases can not be eliminated:
-// "SELECT b, a FROM t",
-// or "SELECT c AS a, c AS b FROM t WHERE d = 1",
-// or "SELECT t1.a, t2.b, t1.b, t2.a FROM t1, t2 WHERE t1.a < 0 AND t2.b > 0".
-func projectionCanBeEliminated(p *Projection) bool {
-	child := p.children[0].(PhysicalPlan)
+	return true
+}
+
+// canProjectionBeEliminatedStrict checks whether a projection can be eliminated, returns true if
+// the projection just copy its child's output.
+func canProjectionBeEliminatedStrict(p *Projection) bool {
+	child := p.Children()[0]
 	if p.Schema().Len() != child.Schema().Len() {
 		return false
 	}
 	for i, expr := range p.Exprs {
 		col, ok := expr.(*expression.Column)
-		if !ok {
-			return false
-		}
-		if col.FromID != child.Schema().Columns[i].FromID || col.Position != child.Schema().Columns[i].Position {
+		if !ok || !col.Equal(child.Schema().Columns[i], nil) {
 			return false
 		}
 	}
 	return true
 }
+
+func resolveColumnAndReplace(origin *expression.Column, replace map[string]*expression.Column) {
+	dst := replace[string(origin.HashCode())]
+	if dst != nil {
+		*origin = *dst
+	}
+}
+
+func resolveExprAndReplace(origin expression.Expression, replace map[string]*expression.Column) {
+	switch expr := origin.(type) {
+	case *expression.Column:
+		resolveColumnAndReplace(expr, replace)
+	case *expression.ScalarFunction:
+		for _, arg := range expr.GetArgs() {
+			resolveExprAndReplace(arg, replace)
+		}
+	}
+}
+
+// eliminatePhysicalProjection should be called after physical optimization to eliminate the redundant projection
+// left after logical projection elimination.
+func eliminatePhysicalProjection(p PhysicalPlan, replace map[string]*expression.Column) PhysicalPlan {
+	children := make([]Plan, 0, len(p.Children()))
+	for _, child := range p.Children() {
+		children = append(children, eliminatePhysicalProjection(child.(PhysicalPlan), replace))
+	}
+	p.SetChildren(children...)
+
+	for _, dst := range p.Schema().Columns {
+		resolveColumnAndReplace(dst, replace)
+	}
+	for _, key := range p.Schema().Keys {
+		for _, keyCol := range key {
+			resolveColumnAndReplace(keyCol, replace)
+		}
+	}
+	p.replaceExprColumns(replace)
+
+	proj, isProj := p.(*Projection)
+	if !isProj || !canProjectionBeEliminatedStrict(proj) {
+		return p
+	}
+	child := p.Children()[0]
+	exprs := proj.Exprs
+	childCols := child.Schema().Columns
+	for i, parentColumn := range proj.Schema().Columns {
```

Comment:
No. actually, we only need to refresh child's schema in physical projection elimination when the eliminated projection is root.